From 04fd79a93b094f94ad825c6cce63c139128e89a6 Mon Sep 17 00:00:00 2001
From: Le Hoang Quyen <lehoangquyen@chromium.org>
Date: Wed, 28 Oct 2026 01:22:37 -0700
Subject: [PATCH] [M156] graphite: Remove SK_LEGACY_GRAPHITE_READ_PIXELS_BOTTOM_LEFT_BEHAVIOR

Original change's description:
> graphite: Remove SK_LEGACY_GRAPHITE_READ_PIXELS_BOTTOM_LEFT_BEHAVIOR
>
> Skia now handles bottom-left origin images internally during
> asyncReadPixels. Remove the legacy define and the workarounds
> in CopySharedImageHelper for Graphite readback.
>
> Bug: 553467540
> Change-Id: Iab440ed601ebdcc75ed2265eea6351838a944cab
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8478291
> Reviewed-by: Colin Blundell <blundell@chromium.org>
> Reviewed-by: Michael Ludwig <michaelludwig@google.com>
> Commit-Queue: Quyen Le <lehoangquyen@google.com>
> Cr-Commit-Position: refs/heads/main@{#1707092}

(cherry picked from commit 9e4cb764038c65dc2f2170bb6c1aaf1b5e250814)

Bug: 578459433,553467540
Change-Id: Iab440ed601ebdcc75ed2265eea6351838a944cab
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8671978
Auto-Submit: chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Reviewed-by: Quyen Le <lehoangquyen@google.com>
Reviewed-by: Colin Blundell <blundell@chromium.org>
Commit-Queue: Quyen Le <lehoangquyen@google.com>
Cr-Commit-Position: refs/branch-heads/8078@{#1682}
Cr-Branched-From: c50edc4214d81e4f944c3415d55104a49e46d214-refs/heads/main@{#1706539}
---

--- a/gpu/command_buffer/service/copy_shared_image_helper.cc
+++ b/gpu/command_buffer/service/copy_shared_image_helper.cc
@@ -609,7 +609,6 @@
 // that the src rect is contained within the image.
 bool GraphiteImageReadPixels(GraphiteSharedContext* graphite_shared_context,
                              sk_sp<SkImage> sk_image,
-                             GrSurfaceOrigin src_surface_origin,
                              int src_x,
                              int src_y,
                              const SkImageInfo& dst_info,
@@ -622,23 +621,6 @@
   CHECK(graphite_shared_context);
   ReadPixelsContext context;
 
-#if defined(SK_LEGACY_GRAPHITE_READ_PIXELS_BOTTOM_LEFT_BEHAVIOR)
-  // Make the `src_rect` relative to the bottom left origin if needed. This
-  // works around lack of support for bottom-left origin data during readback in
-  // Graphite. Graphite correctly handles bottom-left origins when rendering, so
-  // if asyncRescaleAndReadPixels() were to render `sk_image` for any reason,
-  // this adjustment won't work. But this call isn't scaling and is presumably
-  // for a shared image that has copy-src usage, so it shouldn't happen. This is
-  // also a no-op when the src rect is the entire image, regardless of origin.
-  if (src_surface_origin == kBottomLeft_GrSurfaceOrigin) {
-    src_y = src_image_bounds.height() - src_rect.y() - src_rect.height();
-    src_rect = gfx::Rect(
-        src_rect.x(), src_y, src_rect.width(), src_rect.height());
-    // This adjustment should not change the fact that src_rect is still valid
-    CHECK(src_image_bounds.Contains(src_rect));
-  }
-#endif
-
   // We don't need to insert a recording since asyncRescaleAndReadPixels is a
   // context operation that inserts its own recording internally.
   if (!graphite_shared_context->asyncRescaleAndReadPixelsAndSubmit(
@@ -657,19 +639,9 @@
     return false;
   }
 
-#if defined(SK_LEGACY_GRAPHITE_READ_PIXELS_BOTTOM_LEFT_BEHAVIOR)
-  // Use CopyPlane to flip as Graphite doesn't support bottom left origin
-  // images. Using a negative height causes CopyPlane to flip while copying.
-  // TODO(crbug.com/40269891): Remove this if Graphite performs the flip
-  // once it supports bottom left origin images.
-  const int height = src_surface_origin == kTopLeft_GrSurfaceOrigin
-                         ? dst_info.height()
-                         : -dst_info.height();
-#else
   // Must copy the async results (often a GPU-mapped buffer) to the CPU
   // pixel_address.
   const int height = dst_info.height();
-#endif
   libyuv::CopyPlane(static_cast<const uint8_t*>(context.async_result->data(0)),
                     context.async_result->rowBytes(0),
                     static_cast<uint8_t*>(pixel_address), row_bytes,
@@ -753,8 +725,8 @@
         shared_context_state_->graphite_shared_context();
     success =
         GraphiteImageReadPixels(graphite_shared_context, std::move(sk_image),
-                                source_shared_image->surface_origin(), src_x,
-                                src_y, dst_info, pixel_address, row_bytes);
+                                src_x, src_y, dst_info, pixel_address,
+                                row_bytes);
   }
   if (!success) {
     return base::unexpected(GLError(GL_INVALID_OPERATION, "ReadPixels",
--- a/skia/config/SkUserConfig.h
+++ b/skia/config/SkUserConfig.h
@@ -147,8 +147,6 @@
 
 #define SK_SUPPORT_LEGACY_GRAPHITE_RRECT_BLUR
 
-#define SK_LEGACY_GRAPHITE_READ_PIXELS_BOTTOM_LEFT_BEHAVIOR
-
 #define SK_DISABLE_GRAPHITE_NONINTEGRAL_VARYINGS_FLAT_INTERPOLATION
 
 ///////////////////////// Imported from BUILD.gn and skia_common.gypi
